### ISSUE TITLE
incorrect gradient for complex matmul

### ISSUE THREAD

**Issue by woodshop** [tensorflow/tensorflow#4859]
The `MatMul` kernel is registered for `complex64` and `complex128` on CPU and GPU. However the gradient is incorrect. This issue has not been caught because there is no test for a complex-valued `MatMul` gradient in `tensorflow\python\kernel_tests\matmul_op_test.py`.

When computing the complex gradient we need to apply the conjugate of the input. Several cwise gradient ops in TF were ignoring this step until it was addressed by @girving and @tensorflower-gardener in https://github.com/tensorflow/tensorflow/commit/821063df9f0e6a0eec8cb78cb0ddc5c5b2b91b9f. However it seems that `MatMul` snuck through the cracks.

**To Reproduce**

``` python
ops.reset_default_graph()
sess = tf.Session()
x = (np.linspace(-3, 3, 6) + 1j*np.linspace(3, -3, 6)).reshape(3, 2)
y = (np.linspace(-3, 3, 8) + 1j*np.linspace(3, -3, 8)).reshape(2, 4)
with sess.as_default():
    with tf.device('/cpu'):
        x_tf = tf.constant(x, dtype=tf.complex64, name="x")
        y_tf = tf.constant(y, dtype=tf.complex64, name="y")
        m = tf.matmul(x_tf, y_tf, name="matmul")
        err = tf.test.compute_gradient_error(x_tf, [3, 2], m, [3, 4])
        print(err)

with sess.as_default():
    with tf.device('/gpu'):
        x_tf = tf.constant(x, dtype=tf.complex64, name="x")
        y_tf = tf.constant(y, dtype=tf.complex64, name="y")
        m = tf.matmul(x_tf, y_tf, name="matmul")
        err = tf.test.compute_gradient_error(x_tf, [3, 2], m, [3, 4])
        print(err)
```

``` python
>>> 6.00002098083
>>> 6.00002098083
```

**Proposed Fix**
Replace [these lines](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L697-L714) with:

``` python
@ops.RegisterGradient("MatMul")
def _MatMulGrad(op, grad):
  with ops.control_dependencies([grad.op]):
      inp0 = math_ops.conj(op.inputs[0])
      inp1 = math_ops.conj(op.inputs[1])
      t_a = op.get_attr("transpose_a")
      t_b = op.get_attr("transpose_b")
      if not t_a and not t_b:
        return (math_ops.matmul(grad, inp1, transpose_b=True),
                math_ops.matmul(inp0, grad, transpose_a=True))
      elif not t_a and t_b:
        return (math_ops.matmul(grad, inp1),
                math_ops.matmul(grad, inp0, transpose_a=True))
      elif t_a and not t_b:
        return (math_ops.matmul(op.inp1, grad, transpose_b=True),
                math_ops.matmul(op.inp0, grad))
      elif t_a and t_b:
        return (math_ops.matmul(op.inp1, grad, transpose_a=True,
                                transpose_b=True),
                math_ops.matmul(grad, op.inp0, transpose_a=True,
                                transpose_b=True))
```

and add a complex gradient test to `tensorflow\python\kernel_tests\matmul_op_test.py`.

**Comment by woodshop** [comment]
I'm not sure I am thinking about this correctly. I've noticed from a previous issue that I posted (https://github.com/tensorflow/tensorflow/issues/2516) that the numerical and theoretical gradients provided by `tf.test.compute_gradient_error` should not be the same in the case of complex ops. I'll close this issue and reopen later if necessary.

**Comment by woodshop** [comment]
I've considered my question a bit more and I believe this issue should remain open. I would love to hear someone weigh in (perhaps @ibab, who has been involved in resolving several TF issues for complex computations).

As noted above, the recent commit  https://github.com/tensorflow/tensorflow/commit/821063df9f0e6a0eec8cb78cb0ddc5c5b2b91b9f by @girving and @@tensorflower-gardener changed the python-computed gradients for several complex analytic functions. I believe that `tf.matmul` needs a similar upgrade. As evidence, TF returns different gradients for scalar `matmul` and scalar `mul`, so one of them must be incorrect, right?

```python
sess = tf.Session()
x = (np.random.randn(1,1) + 1j*np.random.randn(1,1)).astype(np.complex64)
y = (np.random.randn(1,1) + 1j* np.random.randn(1,1)).astype(np.complex64)
with sess.as_default():
    x_tf = tf.constant(x)
    y_tf = tf.constant(y) 
    f_mul = x_tf * y_tf
    f_matmul = tf.matmul(x_tf, y_tf)
    assert f_mul.eval() == f_matmul.eval()
    print('df/dx computed using tf.mul:\n\t{}'.format(sess.run(tf.gradients(f_mul, x_tf))))
    print('df/dx computed using tf.matmul:\n\t{}'.format(sess.run(tf.gradients(f_matmul, x_tf))))
```
```python
>>> df/dx computed using tf.mul:
    	[array([[ 2.16470551+0.75122446j]], dtype=complex64)]
>>> df/dx computed using tf.matmul:
	[array([[ 2.16470551-0.75122446j]], dtype=complex64)]
```

**Comment by jart** [comment]
@woodshop Thank you for putting the effort into creating a very readable issue.

@girving Could you please take a look into this one?

**Comment by girving** [comment]
@jart On leave at the moment.

**Comment by jart** [comment]
@ebrevdo can you take a look on this since @girving is not available right now?

**Comment by woodshop** [comment]
As a follow-up: I confirmed empirically that the complex gradient is incorrect and I have been using `graph.gradient_override_map` to successfully train my models. I'll be happy to create a pull request with the solution proposed above if a googler agrees.

**Comment by woodshop** [comment]
It appears that this issue was recently fixed by 3fe917bfb116ecebdd73de632c18cef6a1a04d9c. I'll leave it open until a googler confirms.

**Comment by concretevitamin** [comment]
I believe the referred commit does fix the issue.
